vllm/v1/core/kv_cache_manager.py

```diff
@@ -278,6 +278,54 @@ def free(self, request: Request) -> None:
             if block.ref_cnt == 0:
                 self.free_block_queue.append(block)
 
+    def get_num_common_prefix_blocks(
+        self,
+        request: Request,
+        num_running_requests: int,
+    ) -> int:
+        """Calculate the number of common prefix blocks shared by all requests
+        in the RUNNING state.
+
+        The function determines this by selecting any request and iterating
+        through its blocks.  A block is considered a common prefix block if its
+        `ref_cnt` equals the total number of requests in the RUNNING state.
+
+        NOTE(woosuk): The number of requests in the RUNNING state is **greater
+        than or equal to** the number of requests scheduled in the current step.
+        This is because the RUNNING state only indicates that:
+        1. The request has not yet finished, and 
+        2. The request holds its blocks unfreed.
+
+        While all scheduled requests must be in the RUNNING state, the inverse
+        is not necessarily true. There may be RUNNING requests that are not
+        scheduled in the current step. 
```

Comment:
Could you elaborate on this? In what case a RUNNING request won't be scheduled in the current step? My understanding is if we cannot schedule all requests in the RUNNING queue, then we will preempt some of them. Preemption frees all their blocks and puts them back to the WAITING queue.

Also we have the following assertion. Doesn't it guarantee the number of scheduled requests is always equal to the number of RUNNING requests?

```
        assert (len(scheduled_new_reqs) + len(scheduled_resumed_reqs) +
                len(scheduled_running_reqs) == len(self.running))
```

Reply:
@comaniac You're right. In our current scheduler, all requests in the RUNNING state are scheduled every step. In other words, RUNNING requests are either scheduled or preempted (with the PREEMPTED state).

I wrote this comment because this might not be the case in the (near) future, where we would allow more flexibility in scheduling. I added the following to the docstring:
>  As of 1/1/2025, the scheduler does not allow this case, but it is possible in the future, as we allow more flexible scheduling.